oci/layer/utils.go

```diff
@@ -229,11 +230,18 @@ func InnerErrno(err error) error {
 
 // isOverlayWhiteout returns true if the FileInfo represents an overlayfs style
 // whiteout (i.e. mknod c 0 0) and false otherwise.
-func isOverlayWhiteout(info os.FileInfo) bool {
-	sysStat := info.Sys().(*unix.Stat_t)
-	if unix.Major(uint64(sysStat.Rdev)) != 0 || unix.Minor(uint64(sysStat.Rdev)) != 0 {
-		return false
+func isOverlayWhiteout(info os.FileInfo) (bool, error) {
+	if unixStat, ok := info.Sys().(*unix.Stat_t); ok {
+		if unix.Major(uint64(unixStat.Rdev)) != 0 || unix.Minor(uint64(unixStat.Rdev)) != 0 {
+			return false, nil
+		}
+	} else if syscallStat, ok := info.Sys().(*syscall.Stat_t); ok {
+		if unix.Major(uint64(syscallStat.Rdev)) != 0 || unix.Minor(uint64(syscallStat.Rdev)) != 0 {
+			return false, nil
+		}
+	} else {
```

Comment:
:grimacing: Oh boy, I was hoping we wouldn't hit this (I thought that `unix.Stat_t` was a type alias). This might be a slightly nicer way of handling it:

```go
func isOverlayWhiteout(info os.FileInfo) bool {
	var major, minor uint32
	switch stat := info.Sys().(type) {
	case *unix.Stat_t:
		major = unix.Major(stat.Rdev)
		minor = unix.Minor(stat.Rdev)
	case *syscall.Stat_t:
		major = unix.Major(stat.Rdev)
		minor = unix.Minor(stat.Rdev)
	default:
		panic("unknown info type")
	}
	return major == 0 && minor == 0 &&
		info.Mode()&os.ModeCharDevice == os.ModeCharDevice
}
```

Though arguably, we should only ever handle `syscall.Stat_t` since I think `os.FileInfo` will only ever use the syscall types. But :shrug:.

Reply:
This used to work when I tested it, so I wonder if they weren't type alises before, but now they're not with newer unixes? Anyway, yeah, I can switch it to this. I don't want to panic() though, seems impolite to panic from libraries.

Reply:
Yeah, switch it to an error if you like (though in principle this is actually a programmer error which is /in theory/ when Go libraries should panic, but I don't really mind :shrug:).